### PR TITLE
restrict env variables start with HASURA_GRAPHQL_ for headers configuration in actions, event triggers & remote schemas

### PR DESCRIPTION
**Pull request by rakeshkky**
### Description

In `headers` definition for actions, remote schemas & event triggers, don't allow `value_from_env` values start with `HASURA_GRAPHQL_`. 

### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components

- [x] Server
- [x] Tests

### Related Issues

N/A

### Steps to test and verify

Creating action/event trigger having headers configuration with `HASURA_GRAPHQL_*` env variables.

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [x] No
- [ ] Yes

#### Metadata

Does this PR add a new Metadata feature?
- [x] No
- [ ] Yes

#### GraphQL
- [x] No new GraphQL schema is generated
- [ ] New GraphQL schema is being generated

#### Breaking changes

- [ ] No Breaking changes
- [x] There are breaking changes:

  1. Metadata API

     Existing `query` types:
     - [ ] Modify `args` payload which is not backward compatible
     - [x] Behavioural change of the API
     - [ ] Change in response `JSON` schema
     - [ ] Change in error code
     

If metadata contains any action/event trigger/remote schema defined with `HASURA_GRAPHQL_*` env variables in headers configuration, then such graphql-engine isn't upgrade-able. To upgrade, recreate those actions/event triggers/remote schemas without restricted env variables.

### CODE REVIEW

**Comment by tirumaraiselvan** [comment]
@rakeshkky Remote schemas as well?

**Comment by rakeshkky** [comment]
> @rakeshkky Remote schemas as well?

Yes @tirumaraiselvan . Updating the title accordingly.

File: server/src-lib/Hasura/RQL/DDL/Headers.hs [hunk 1]
```diff
@@ -35,7 +35,10 @@ instance FromJSON HeaderConf where
     case (value, valueFromEnv ) of
       (Nothing, Nothing)  -> fail "expecting value or value_from_env keys"
       (Just val, Nothing) -> return $ HeaderConf name (HVValue val)
-      (Nothing, Just val) -> return $ HeaderConf name (HVEnv val)
+      (Nothing, Just val) -> do
+        when (T.isPrefixOf "HASURA_GRAPHQL_" val) $
+          fail "env variables start with \"HASURA_GRAPHQL_\" are not allowed for value_from_env"
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/DDL/Headers.hs:40]:
```suggestion
          fail "env variables starting with \"HASURA_GRAPHQL_\" are not allowed in value_from_env"
```

File: server/tests-py/queries/actions/metadata/create_with_headers.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      - name: name
+        type: String!
+      output_type: User!
+      handler: http://127.0.0.1:5593/create-user
+      headers:
+      - name: x-client-id
+        value_from_env: HASURA_GRAPHQL_CLIENT_NAME
+response:
+  path: $.definition.headers[0]
+  error: env variables start with "HASURA_GRAPHQL_" are not allowed for value_from_env
```
Comment (0x777) [inline server/tests-py/queries/actions/metadata/create_with_headers.yaml:20]:
```suggestion
  error: env variables starting with "HASURA_GRAPHQL_" are not allowed in value_from_env: HASURA_GRAPHQL_CLIENT_NAME
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,10 +2,20 @@
 
 ## Next release
 
+### Breaking change
+
+Environment variables starting with `HASURA_GRAPHQL_` are restricted to configure `value_from_env` key
```
Comment (0x777) [inline CHANGELOG.md:7]: @tirumaraiselvan Please to review the changelog

**Comment by 0x777** [comment]
@rakeshkky Looks like the changes made to the test case results in an invalid yaml error.

**Comment by ItsWendell** [comment]
What was the reason this was implemented? I've been using passing `HASURA_GRAPHQL_ADMIN_SECRET` to my serverless API functions that in turn check if the secret matches with their environment variables to ensure a request is coming from hasura. In a new version I've separated these secrets into separate variables as a sort of API 'key'.

Was this an security issue?

**Comment by perrosnk** [comment]
What was the reason for that? I have been doing the same thing as @ItsWendell

**Comment by greatwitenorth** [comment]
I had the same use case as @ItsWendell. Can someone from the Hasura team let us know the best practice now when dealing with serverless functions and trying to authenticate the validity of the request?

**Comment by tirumaraiselvan** [comment]
The main reason was that people with limited access to console  (in Hasura Cloud/Enterprise) could see such environment variables in the event trigger logs. 

> I've been using passing HASURA_GRAPHQL_ADMIN_SECRET to my serverless API functions that in turn check if the secret matches with their environment variables to ensure a request is coming from hasura. In a new version I've separated these secrets into separate variables as a sort of API 'key'.

We understand this may be inconvenient for your setup but it may be good practice to use a secret which has lower impact (i.e. only used to ascertain origin) for this use-case.

**Pull request merged**